webapp/src/ts/services/enketo.service.ts

```diff
@@ -447,11 +451,34 @@ export class EnketoService {
       e._couchId = id;
     };
 
+    mapOrAssignId($record[0], doc._id || uuid());
+
     const getId = (xpath) => {
-      return recordDoc
-        .evaluate(xpath, recordDoc, null, XPathResult.ANY_TYPE, null)
-        .iterateNext()
-        ._couchId;
+      const xPathResult = recordDoc.evaluate(xpath, recordDoc, null, XPathResult.ANY_TYPE, null);
+      let node = xPathResult.iterateNext();
+      while (node) {
+        if (node._couchId) {
+          return node._couchId;
+        }
+        node = xPathResult.iterateNext();
+      }
+    };
+
+    const getClosestPath = (element, $element, path) => {
+      path = path.trim();
+      const relativeReference = path.startsWith('./');
+      const repeatReference = repeatPaths?.find(repeatPath => path.startsWith(repeatPath));
+
+      if (!relativeReference && !repeatReference) {
+        return;
+      }
+
+      // assign a unique id for xpath context, since the element can be inside a repeat
+      element.id = element.id || uuid();
+      const uniqueElementSelector = `${element.nodeName}[@id="${element.id}"]`;
+      const localPath = path.replace(repeatReference ? `${repeatReference}/` : './', '');
+
+      return `//${uniqueElementSelector}/ancestor-or-self::*/descendant-or-self::${localPath}`;
     };
```

Comment:
As far as I can tell this is calculating two separate components of the path, but it's all jumbled up together which makes it harder to read than it needs to be. What about separating the relative path calculation from the unique element selector, eg...

```suggestion
    const getRelativePath = (path) => {
      const repeatReference = repeatPaths?.find(repeatPath => path.startsWith(repeatPath));
      if (repeatReference) {
        return path.replace(`${repeatReference}/`, '');
      }
      if (path.startsWith('./')) {
        return path.replace(`./`, '');
      }
      return;
    };

    const getClosestPath = (element, $element, path) => {
      const relativePath = getRelativePath(path.trim());
      if (!relativePath) {
        return;
      }

      // assign a unique id for xpath context, since the element can be inside a repeat
      if (!element.id) {
        element.id = uuid();
      }
      const uniqueElementSelector = `${element.nodeName}[@id="${element.id}"]`;

      return `//${uniqueElementSelector}/ancestor-or-self::*/descendant-or-self::${relativePath}`;
    };
```